test/integration/scheduler/queue_test.go

```diff
@@ -178,89 +179,185 @@ func TestSchedulingGates(t *testing.T) {
-		pendingPods, _ := testCtx.Scheduler.SchedulingQueue.PendingPods()
-		return len(pendingPods) == 3, nil
-	}); err != nil {
-		t.Fatal(err)
-	}
+				if tt.initialPod != nil {
+					if _, err := cs.CoreV1().Pods(ns).Create(ctx, tt.initialPod, metav1.CreateOptions{}); err != nil {
+						t.Fatalf("Failed to create an initial Pod %q: %v", tt.initialPod.Name, err)
+					}
+
+					// Wait for the pod to be present in the scheduling queue.
+					if err := wait.PollUntilContextTimeout(ctx, time.Millisecond*200, wait.ForeverTestTimeout, false, func(ctx context.Context) (bool, error) {
+						pendingPods, _ := testCtx.Scheduler.SchedulingQueue.PendingPods()
+						return len(pendingPods) == 1, nil
+					}); err != nil {
+						t.Fatal(err)
+					}
+					scheduleNextPod(t, testCtx, true /* this Pod is supposed to be scheduled */)
+				}
 
-	// Pop the three pods out. They should be unschedulable.
-	for i := 0; i < 3; i++ {
-		podInfo := testutils.NextPodOrDie(t, testCtx)
-		fwk, ok := testCtx.Scheduler.Profiles[podInfo.Pod.Spec.SchedulerName]
-		if !ok {
-			t.Fatalf("Cannot find the profile for Pod %v", podInfo.Pod.Name)
-		}
-		// Schedule the Pod manually.
-		_, fitError := testCtx.Scheduler.SchedulePod(ctx, fwk, framework.NewCycleState(), podInfo.Pod)
-		if fitError == nil {
-			t.Fatalf("Expect Pod %v to fail at scheduling.", podInfo.Pod.Name)
-		}
-		testCtx.Scheduler.FailureHandler(ctx, fwk, podInfo, framework.NewStatus(framework.Unschedulable).WithError(fitError), nil, time.Now())
-	}
+				for _, pod := range tt.pods {
+					if _, err := cs.CoreV1().Pods(ns).Create(ctx, pod, metav1.CreateOptions{}); err != nil {
+						t.Fatalf("Failed to create Pod %q: %v", pod.Name, err)
+					}
+				}
 
-	// Trigger a NodeTaintChange event.
-	// We expect this event to trigger moving the test Pod from unschedulablePods to activeQ.
-	node.Spec.Taints = nil
-	if _, err := cs.CoreV1().Nodes().Update(ctx, node, metav1.UpdateOptions{}); err != nil {
-		t.Fatalf("Failed to remove taints off the node: %v", err)
+				// Wait for the tt.pods to be present in the scheduling queue.
+				if err := wait.PollUntilContextTimeout(ctx, time.Millisecond*200, wait.ForeverTestTimeout, false, func(ctx context.Context) (bool, error) {
+					pendingPods, _ := testCtx.Scheduler.SchedulingQueue.PendingPods()
+					return len(pendingPods) == len(tt.pods), nil
+				}); err != nil {
+					t.Fatal(err)
+				}
+
+				t.Log("Confirmed Pods in the scheduling queue, starting to schedule them")
+
+				// Pop all pods out. They should be unschedulable.
+				for i := 0; i < len(tt.pods); i++ {
+					scheduleNextPod(t, testCtx, false /* all Pods are supposed to fail at scheduling */)
+				}
+
+				t.Log("finished initial schedulings for all Pods, will trigger triggerFn")
+
+				err := tt.triggerFn(testCtx)
+				if err != nil {
+					t.Fatalf("Failed to trigger the event: %v", err)
+				}
+
+				time.Sleep(1 * time.Second)
```

Comment:
why the sleep?

Reply:
Changed it to wait.PollUntilContextTimeout.